### PR TITLE
Support importing panorama photo

### PR DESCRIPTION
**Pull request by miurahr**
Implement #89

* [Facebook explanation ](https://www.facebook.com/notes/eric-cheng/editing-360-photos-injecting-metadata/10156930564975277)  for 360 degree photo conditions.

> * The photo must have a 2:1 aspect ratio, with maximum file dimensions of 6000 x 3000 pixels
> * The Exif XMP tag, “ProjectionType=equirectangular”

### CODE REVIEW

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java [hunk 1]
```diff
@@ -73,7 +73,10 @@ public static MapillaryImportedImage createInstance(final ImageEntry geoImage) {
     final long time = geoImage.hasGpsTime()
       ? geoImage.getGpsTime().getTime()
       : geoImage.hasExifTime() ? geoImage.getExifTime().getTime() : System.currentTimeMillis();
-    return new MapillaryImportedImage(coord, ca, geoImage.getFile(), time);
+    final double width = geoImage.getWidth();
+    final double height = geoImage.getHeight();
+    final boolean pano = (width >= 4096) && (width == height * 2);
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java]: I'm not familiar enough with Mapillary to know if a panorama has a fixed ratio of `width == height * 2`. Is it always the case, or would `width >= height * 2` be safer?

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageImportUtil.java [hunk 2]
```diff
@@ -113,10 +112,12 @@ private static MapillaryImportedImage readImageFrom(
     } else {
       ca = 0;
     }
+    FileInputStream nis = new FileInputStream(f);
+    boolean pano = ImageMetaDataUtil.getPanorama(nis);
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageImportUtil.java]: shouldn't the FileInputStream be closed?

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    return pano;
+  }
+
+  public static boolean getPanorama(final Reader sr) {
+    boolean pano = false;
+    try {
+      DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+      factory.setNamespaceAware(true);
+      DocumentBuilder builder = factory.newDocumentBuilder();
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java]: Please use XmlUtils.newSafeDOMBuilder()

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        Node rdfNode = xmpMetaNodeList.item(i);
+        if (rdfNode.getNodeType() == Node.ELEMENT_NODE) {
+          Element rdfElement = (Element) rdfNode;
+          NodeList rdfChildNodeList = rdfElement.getChildNodes();
+          for (int j = 0; j < rdfChildNodeList.getLength(); j++) {
+            Node rdfChildNode = rdfChildNodeList.item(j);
+            if (rdfChildNode.getNodeType() == Node.ELEMENT_NODE) {
+              Element rdfChildElement = (Element) rdfChildNode;
+              String projection = rdfChildElement.getAttributeNS("http://ns.google.com/photos/1.0/panorama/", "ProjectionType");
+              if (projection != null && projection.equals("equirectangular")) {
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java]: you can simply use `if ("equirectangular".equals(projection)) {`

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    // private util.
+  }
+
+  public static boolean getPanorama(final InputStream is) {
+    String xmpxml = null;
+    try {
+      xmpxml = Imaging.getXmpXml(is, null);
+    } catch (ImageReadException | IOException ex) {
+      // Can't read XML metadata. use default instead.
+      return false;
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java]: Please log exceptions at least with trace level: `Logging.trace(ex)`

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+              if (projection != null && projection.equals("equirectangular")) {
+                pano = true;
+                break;
+              }
+            }
+          }
+        }
+      }
+    } catch (ParserConfigurationException | SAXException | IOException ex) {
+      // Can't read XML metadata. use default instead.
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java]: Please log exceptions at least with trace level: `Logging.trace(ex)`

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java (hunk 1 above)
Comment (HolgerJeromin) [reply src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java]: A user can upload any image from any camera to Mapillary.

Comment (miurahr) [reply src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java]: 360 degree photo, aka. equirectangular projection,   always become  1:2 ratio of size.
1:2 ratio is not always 360 degree photo in reverse.
It is just a heuristics to consider width == height * 2 as panorama photo. Are their any way to detect 360 degree photo here? If not, we can only simply set `boolean pano  = false`

Comment (miurahr) [reply src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java]: Next push change here to use get panorama()

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageImportUtil.java (hunk 2 above)
Comment (miurahr) [reply src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageImportUtil.java]: Add close()

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.openstreetmap.josm.tools.XmlUtils;
+
+
+public class ImageMetaDataUtil {
+
+  private ImageMetaDataUtil() {
+    // private util.
+  }
+
+  public static boolean getPanorama(final InputStream is) {
```
Comment (miurahr) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java]: It is better to change method name to `isPano` or `isPanorama`

**Comment by miurahr** [comment]
All the change requested are resolved. 
However,  I'd like to make more change and does refactoring on ImageMetaDataUtil.java and ImageMetaDataUtilTest.java to be a more generic to be able to use for other features such as downloaded photos.

**Comment by don-vip** [comment]
@miurahr Thanks! Do you want to do the refactoring in this merge request? I can wait, no problem :)

**Comment by miurahr** [comment]
@don-vip Now iI'm satisfied and want you to merge it.

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java [hunk 8]
```diff
@@ -73,7 +76,13 @@ public static MapillaryImportedImage createInstance(final ImageEntry geoImage) {
     final long time = geoImage.hasGpsTime()
       ? geoImage.getGpsTime().getTime()
       : geoImage.hasExifTime() ? geoImage.getExifTime().getTime() : System.currentTimeMillis();
-    return new MapillaryImportedImage(coord, ca, geoImage.getFile(), time);
+    boolean pano = false;
+    try {
+      pano = ImageMetaDataUtil.isPanorama(new FileInputStream(geoImage.getFile()));
```
Comment (don-vip) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java]: it seems this input stream is not closed

Comment (miurahr) [reply src/main/java/org/openstreetmap/josm/plugins/mapillary/MapillaryImportedImage.java]: Ok, Id like to use ' try (FileInputStream fis = new FileInputStream(brah)) { }' style for here.

**Comment by don-vip** [comment]
Fine for me. @floscher @jpietri do you want to take a look?

File: src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    return pano;
+  }
+
+  /**
+   * check whether image file is a panorama photo or not.
+   * @param image a BufferedImage to check.
+   * @return true if image is a panorama.
+   */
+  public static boolean isPanorama(final BufferedImage image) {
```
Comment (miurahr) [inline src/main/java/org/openstreetmap/josm/plugins/mapillary/utils/ImageMetaDataUtil.java]: I found BufferedImage class don't hold a meta data. I'd like to remove this method signature.

**Comment by miurahr** [comment]
Any Commnents for panorama improvements? @floscher @jpietri

**Comment by jpietri** [comment]
Hello @miurahr, thanks for this contribution and sorry for this late answer.
About to merge your PR.
Still, while testing it, I've injected the param `projectionType` with www.thexifer.net, and it ended up as an XML element, and not an attribute. As I know nothing about 360 images, I've checked a bit on the Internet, and I've seen it more often as an element.
Will create a PR with some additional code to grab the element and let you review it...

**Pull request merged**